packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.Input/src/Extensions/StringExtensions.cs

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.TypeSpec.Generator.Input.Extensions
+{
+    public static class StringExtensions
+    {
+        private static bool IsWordSeparator(char c) => !SyntaxFacts.IsIdentifierPartCharacter(c) || c == '_';
+
+        [return: NotNullIfNotNull("name")]
+        public static string ToIdentifierName(this string name, bool isCamelCase = true)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                return name;
+            }
+            StringBuilder nameBuilder = new StringBuilder();
+
+            int i = 0;
+
+            if (char.IsDigit(name[0]))
+            {
+                nameBuilder.Append("_");
+            }
+            else
+            {
+                while (!SyntaxFacts.IsIdentifierStartCharacter(name[i]))
+                {
+                    i++;
+                }
+            }
+
+            bool upperCase = false;
+            int firstWordLength = 1;
+            for (; i < name.Length; i++)
+            {
+                var c = name[i];
+                if (IsWordSeparator(c))
+                {
+                    upperCase = true;
+                    continue;
+                }
+
+                if (nameBuilder.Length == 0 && isCamelCase)
+                {
+                    c = char.ToUpper(c);
+                    upperCase = false;
+                }
+                else if (nameBuilder.Length < firstWordLength && !isCamelCase)
+                {
+                    c = char.ToLower(c);
+                    upperCase = false;
+                    // grow the first word length when this letter follows by two other upper case letters
+                    // this happens in OSProfile, where OS is the first word
+                    if (i + 2 < name.Length && char.IsUpper(name[i + 1]) && (char.IsUpper(name[i + 2]) || IsWordSeparator(name[i + 2])))
+                        firstWordLength++;
+                    // grow the first word length when this letter follows by another upper case letter and an end of the string
+                    // this happens when the string only has one word, like OS, DNS
+                    if (i + 2 == name.Length && char.IsUpper(name[i + 1]))
+                        firstWordLength++;
+                }
+
+                if (upperCase)
+                {
+                    c = char.ToUpper(c);
+                    upperCase = false;
+                }
+
+                nameBuilder.Append(c);
+            }
+
+            return nameBuilder.ToString();
+        }
+
+        [return: NotNullIfNotNull(nameof(name))]
+        public static string ToVariableName(this string name) => name.ToIdentifierName(isCamelCase: false);
+
+        public static string RemovePeriods(this string input)
```

Comment:
Does this need to be public?

Reply:
used in https://github.com/microsoft/typespec/blob/ef9633172e16029e8c75945db47d2413982854a3/packages/http-client-csharp/generator/Microsoft.TypeSpec.Generator.ClientModel/src/Providers/ModelReaderWriterContextDefinition.cs#L13

Reply:
That is the only place it is used - we can just move it into that class.